src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/StackSpiller.Temps.cs

```diff
@@ -205,6 +207,63 @@ private void EnsureDone()
                 }
             }
 
+            private static bool ShouldSaveToTemp(Expression expression)
+            {
+                if (expression == null)
+                    return false;
+
+                // Some expressions have no side-effects and don't have to be
+                // stored into temporaries, e.g.
+                //
+                //     xs[0] = try { ... }
+                //           |
+                //           v
+                //        t0 = xs
+                //        t1 = 0            // <-- this is redundant
+                //        t2 = try { ... }
+                //    t0[t1] = t2
+                //           |
+                //           v
+                //        t0 = xs
+                //        t1 = try { ... }
+                //     t0[0] = t1
+
+                switch (expression.NodeType)
+                {
+                    // Emits ldnull, ldc, initobj, closure constant access, etc.
+                    case ExpressionType.Constant:
+                    case ExpressionType.Default:
+                        return false;
+
+                    // Emits calls to pure RuntimeOps methods with immutable arguments
+                    case ExpressionType.Quote:
+                    case ExpressionType.RuntimeVariables:
+                        return false;
```

Comment:
NB: The rationale for these is that both perform any of the following pure steps:

* Load a constant from the `Closure.Constants` collection; these entries are immutable:
  * A `long[]` for indexes used by `RuntimeVariables` to traverse the parent chain and index into `HoistedLocals`
  * An `Expression` containing an expression tree being `Quote`'d
* Call a runtime helper method in `RuntimeOps`, all of which are pure and are not supposed to throw:
  * Allocation of a `RuntimeVariables` object by simply performing field stores
  * An expression visit for an expression being quoted, at most binding unbound parameters

The `Quote` case could be scrutinized though... The `ExpressionQuoter` will reduce `Extension` nodes which calls into `Reduce` methods. If such a method throws, the timing of the exception changes when spilling is carried out. As far as I know, there's no documentation stating that `Reduce` should not throw, so we may be better off removing the `Quote` case altogether. Let me know what you think.